src/request/body.rs

```diff
@@ -0,0 +1,77 @@
+use std::convert::TryInto;
+use std::fs;
+use std::io::{copy, Result as IoResult, Seek, SeekFrom, Write};
+
+/// The kinds of request bodies currently supported by this crate.
+#[derive(Debug, Clone, Copy)]
+pub enum BodyKind {
+    /// An empty request body
+    Empty,
+    /// A request body with a known length
+    KnownLength(u64),
+}
+
+/// A generic rewindable request body
+pub trait Body {
+    /// Determine the kind of the request body
+    fn kind(&mut self) -> IoResult<BodyKind>;
+
+    /// Write out the request body into the given writer
+    fn write<W: Write>(&mut self, writer: W) -> IoResult<()>;
+}
+
+/// An empty request body
+#[derive(Debug, Clone, Copy)]
+pub struct Empty;
+
+impl Body for Empty {
+    fn kind(&mut self) -> IoResult<BodyKind> {
+        Ok(BodyKind::Empty)
+    }
+
+    fn write<W: Write>(&mut self, _writer: W) -> IoResult<()> {
+        Ok(())
+    }
+}
+
+/// A request body containing UTF-8-encoded text
+#[derive(Debug, Clone)]
+pub struct Text<B>(pub B);
+
+impl<B: AsRef<str>> Body for Text<B> {
+    fn kind(&mut self) -> IoResult<BodyKind> {
+        Ok(BodyKind::KnownLength(self.0.as_ref().len().try_into().unwrap()))
```

Comment:
I'm a bit puzzled by the fact that there no `From<usize> for u64`. `usize` is either 32 or 64 bits and `u64` is always 64 bits... It should be possible.

Reply:
There is no such implementation which future-proofs Rust for 128-bit point width targets, but the current implementation is basically
```
#[cfg(target_pointer_width = "64")]
try_from_unbounded!(usize, u64, u128);
```
and hence will never fail.  So while this is quite verbose, there is no associated runtime overhead.